src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseTestCaseService.java

```diff
@@ -0,0 +1,158 @@
+        this.testCaseRepository = testCaseRepository;
+    }
+
+    // TODO: Workaround for known bug, should be removed once fixed: https://jira.spring.io/browse/DATAJPA-1357
+    // The issue is that these methods are called from the build result notification of bamboo, so there is no authentication object available.
+    // This doesn't cause problems for out-of-the-box repository methods, but does for custom ones.
+    private Authentication getAuthDummy() {
+        return new Authentication() {
+
+            @Override
+            public Collection<? extends GrantedAuthority> getAuthorities() {
+                return null;
+            }
+
+            @Override
+            public Object getCredentials() {
+                return null;
+            }
+
+            @Override
+            public Object getDetails() {
+                return null;
+            }
+
+            @Override
+            public Object getPrincipal() {
+                return null;
+            }
+
+            @Override
+            public boolean isAuthenticated() {
+                return false;
+            }
+
+            @Override
+            public void setAuthenticated(boolean isAuthenticated) throws IllegalArgumentException {
+
+            }
+
+            @Override
+            public String getName() {
+                return null;
+            }
+        };
+    }
+
+    /**
+     * Returns all test cases for a programming exercise.
+     * 
+     * @param id of a programming exercise.
+     * @return test cases of a programming exercise.
+     */
+    public Set<ProgrammingExerciseTestCase> findByExerciseId(Long id) {
+        SecurityContext context = SecurityContextHolder.getContext();
+        context.setAuthentication(getAuthDummy());
+        return this.testCaseRepository.findByExerciseId(id);
+    }
+
+    /**
+     * Returns all active test cases for a programming exercise. Only active test cases are evaluated on build runs.
+     * 
+     * @param id of a programming exercise.
+     * @return active test cases of a programming exercise.
+     */
+    public Set<ProgrammingExerciseTestCase> findActiveByExerciseId(Long id) {
+        SecurityContext context = SecurityContextHolder.getContext();
+        context.setAuthentication(getAuthDummy());
+        return this.testCaseRepository.findActiveByExerciseId(id);
+    }
+
+    /**
+     * From a list of build run feedback, extract all test cases. If an already stored test case is not found anymore in the build result, it will not be deleted, but set inactive.
+     * This way old test cases are not lost, some interfaces in the client might need this information to e.g. show warnings. Returns true if the test cases have changed, false if
+     * they haven't.
+     * 
+     * @param feedbacks list of build log output.
+     * @param exercise  programming exercise.
+     */
+    public boolean generateFromFeedbacks(List<Feedback> feedbacks, ProgrammingExercise exercise) {
```

Comment:
Can we give it a better name? why does it return a boolean?
Please also mention the return value in the JavaDoc

Reply:
- I renamed it to `generateTestCasesFromFeedbacks`
- I placed the return value documentation under `@return`

It returns true if the test cases have changed, if there is no change false.
This information is valuable because I use it to decide if the user on the client should be notified about new test cases.
I could move this action into the method, but I didn't want to have a side effect in it.